pkg/controller/pkg/resolver/reconciler.go

```diff
@@ -0,0 +1,283 @@
+		Owns(&v1beta1.ProviderRevision{}).
+		Complete(r)
+}
+
+// NewReconciler creates a new package revision reconciler.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
+	r := &Reconciler{
+		client:  mgr.GetClient(),
+		lock:    resource.NewAPIFinalizer(mgr.GetClient(), finalizer),
+		log:     logging.NewNopLogger(),
+		record:  event.NewNopRecorder(),
+		newDag:  dag.NewMapDag,
+		fetcher: xpkg.NewNopFetcher(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+
+	return r
+}
+
+// Reconcile package revision.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	lock := &v1alpha1.Lock{}
+	if err := r.client.Get(ctx, req.NamespacedName, lock); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		log.Debug(errGetLock, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetLock)
+	}
+
+	// NOTE(hasheddan): we add a finalizer and never remove it to guard against
+	// a user accidentally deleting the lock, which should never be done.
+	// Consider instead recreating the lock if it is not present.
+	if err := r.lock.AddFinalizer(ctx, lock); err != nil {
+		log.Debug(errAddFinalizer, "error", err)
+		return reconcile.Result{RequeueAfter: shortWait}, nil
+	}
+
+	log = log.WithValues(
+		"uid", lock.GetUID(),
+		"version", lock.GetResourceVersion(),
+		"name", lock.GetName(),
+	)
+
+	dag := r.newDag()
+	implied, err := dag.Init(v1alpha1.ToNodes(lock.Packages))
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, errBuildDAG)
+	}
+
+	// Make sure we don't have any cyclical imports. If we do, refuse to install
+	// additional packages.
+	_, err = dag.Sort()
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, errSortDAG)
+	}
+
+	// If we are missing a node, we want to create it. We only create the first
+	// implied node as we will be requeued when it adds itself to the lock, at
+	// which point we will check for missing nodes again.
+	if len(implied) > 0 {
+		dep, ok := implied[0].(*v1alpha1.Dependency)
+		if !ok {
+			log.Debug(errInvalidDependency, "error", errors.Errorf(errMissingDependencyFmt, dep.Identifier()))
+			return reconcile.Result{}, nil
+		}
+		c, err := semver.NewConstraint(dep.Constraints)
+		if err != nil {
+			log.Debug(errInvalidConstraint, "error", err)
+			return reconcile.Result{}, nil
+		}
+		ref, err := name.ParseReference(dep.Package)
+		if err != nil {
+			log.Debug(errInvalidDependency, "error", err)
+			return reconcile.Result{}, nil
+		}
+
+		// NOTE(hasheddan): we will be unable to fetch tags for private
+		// dependencies because we do not attach any secrets. Consider copying
+		// secrets from parent dependencies.
```

Comment:
Note that this iteration constrains dependency resolution to packages that can be fetched anonymously. Propagating secrets from a parent node to its dependencies will be addressed in a follow-up.